### PR TITLE
api server exception management and code optimization (#397)

### PR DESCRIPTION
**Pull request by hgaol**
## *Tips*
- *Thanks very much for contributing to Apache DolphinScheduler.*
- *Please review https://dolphinscheduler.apache.org/en-us/community/index.html before opening a pull request.*

## What is the purpose of the pull request

Currently, there are many `try-catch` code in `api-server`'s controller. This pr clean the code, use an annotation to point out which `Status` should be return if there is an exception.

I just update `AccessTokenController` firstly, to avoid conflict. If this PR is accepted, maybe I could update for other controllers :)

i.e.
Before
```java
public Result generateToken(@RequestAttribute(value = Constants.SESSION_USER) User loginUser,
                              @RequestParam(value = "userId") int userId,
                              @RequestParam(value = "expireTime") String expireTime){
    logger.info("login user {}, generate token , userId : {} , token expire time : {}",loginUser,userId,expireTime);
    try {
        Map<String, Object> result = accessTokenService.generateToken(userId, expireTime);
        return returnDataList(result);
    }catch (Exception e){
        logger.error(GENERATE_TOKEN_ERROR.getMsg(),e);
        return error(GENERATE_TOKEN_ERROR.getCode(), GENERATE_TOKEN_ERROR.getMsg());
    }
}
```

After
```java
@ControllerException(GENERATE_TOKEN_ERROR)
public Result generateToken(@RequestAttribute(value = Constants.SESSION_USER) User loginUser,
                            @RequestParam(value = "userId") int userId,
                            @RequestParam(value = "expireTime") String expireTime) {
    logger.info("login user {}, generate token , userId : {} , token expire time : {}", loginUser, userId, expireTime);

    Map<String, Object> result = accessTokenService.generateToken(userId, expireTime);
    return returnDataList(result);
}
```

## Brief change log

  - Add `ControllerException` annotation in `org.apache.dolphinscheduler.api.exceptions`.
  - Add `ApiExceptionHandler` controller advice to deal with exceptions.
  - Replace `AccessTokenController` functions which has `try-catch` style with `@ControllerException` annotation.

## Verify this pull request

I've debugged the code locally, and it works well. But it's difficult to write tests for this code, because I should find how to throw an exception. It will be easy if I can write some test controllers to test this PR. Pls tell me if I can do that, or is there any other way for testing.

Thanks in advance!

Hope it has no conflict.

### CODE REVIEW

**Comment by hgaol** [comment]
Hi, is there anyone can help me on the unit test issue? How could I do to solve it... To this situation.

**Comment by dailidong** [comment]
good job， the code becomes more concise. I think you can refer the QueueControllerTest , about Exception Test , refer the following code:
 try {
        mockMvc.perform(post("/api/athlete").contentType(contentType).
                content(TestUtil.convertObjectToJsonBytes(wAthleteFTP)))
                .andExpect(status().isInternalServerError())
                .andExpect(content().string("..."))
                .andDo(print());
    } catch (Exception e){
        //sink it
    }

I want to say I'm not very familar with MockMvc, you can contact me by mail( lidongdai@apache.org )  or    add my wechat(510570367),  when added , please tell me your github id is  hgaol

**Comment by hgaol** [comment]
> good job， the code becomes more concise. I think you can refer the QueueControllerTest , about Exception Test , refer the following code:
> try {
> mockMvc.perform(post("/api/athlete").contentType(contentType).
> content(TestUtil.convertObjectToJsonBytes(wAthleteFTP)))
> .andExpect(status().isInternalServerError())
> .andExpect(content().string("..."))
> .andDo(print());
> } catch (Exception e){
> //sink it
> }
> 
> I want to say I'm not very familar with MockMvc, you can contact me by mail( [lidongdai@apache.org](mailto:lidongdai@apache.org) ) or add my wechat(510570367), when added , please tell me your github id is hgaol

OK, let me have a try. thank you!

**Comment by hgaol** [comment]
I've added some tests, but seems it doesn't run. Does anyone has some ideas on that?

**Comment by hgaol** [comment]
At last, it works, thanks @dailidong for your help.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/AccessTokenService.java [hunk 1]
```diff
@@ -83,6 +84,9 @@
     public Map<String, Object> createToken(int userId, String expireTime, String token) {
         Map<String, Object> result = new HashMap<>(5);
 
+        if (userId <= 0) {
+            throw new ApiServerException("User id should not less than or equals to 0.");
```
Comment (Technoboy-) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/AccessTokenService.java]: Using IllegalArgumentException

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/exceptions/ApiServerException.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.dolphinscheduler.api.exceptions;
+
+/**
+ * ApiException
+ */
+public class ApiServerException extends RuntimeException {
```
Comment (Technoboy-) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/exceptions/ApiServerException.java]: delete this class

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/exceptions/ControllerException.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * controller exception annotation
+ */
+@Retention(RUNTIME)
+@Target(METHOD)
+public @interface ControllerException {
```
Comment (Technoboy-) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/exceptions/ControllerException.java]: How about ApiException.  for you call handler ApiExceptionHandler

**Comment by dailidong** [comment]
good job. thanks for your first contribution, looking forward for your next conttribution, haha

**Pull request merged**